DDI-456 Completed draft of Databricks Import

markzegarelli opened the pull request.

bh202205 commented on docs/data/sources/databricks.md:
```diff
@@ -0,0 +1,113 @@
+
+## Restrictions
+
+- Databricks import doesn't support data well...
```
Are you trying to convey that we don't support data export in this feature? If so, can we just say something like we don't support export data from Amplitude to your Databricks warehouses yet?

markzegarelli replied:
Yes, thanks. This is related the the question I posed in the GDoc yesterday. Will update now.

markzegarelli commented on docs/data/sources/databricks.md:
```diff
@@ -0,0 +1,113 @@
+
+## Restrictions
+
+- Databricks import doesn't support data well...
```
```suggestion
- This integration supports importing data from Databricks.
```

markzegarelli commented on docs/data/sources/databricks.md:
```diff
@@ -0,0 +1,113 @@
+  - Amplitude ignores rows with `delete` operations for all data types.
+- Create joins to import data from multiple tables.
+
+## Restrictions
```
```suggestion
## Limitations
```

markzegarelli commented:
@bh202205 Is there anything else required for this?

bh202205 commented on docs/data/sources/databricks.md:
```diff
@@ -0,0 +1,121 @@
+title: Databricks Import
+---
+
+Amplitude's Databricks import source enables you to import data from Databricks to your Amplitude account. Databricks import uses the Databricks Change Data Feed feature to securely access and extract live data from your Databricks workspace.
```
Can you provide a link to CDF for our customers so they can relate to this technology directly in databricks?
https://docs.databricks.com/en/delta/delta-change-data-feed.html#use-delta-lake-change-data-feed-on-databricks

bh202205 commented on docs/data/sources/databricks.md:
```diff
@@ -0,0 +1,121 @@
+
+## Limitations
+
+- Databricks import doesn't support data well.
```
Let's revise this to be simply `Databricks import doesn't support export data to your own destinations`. I feel like it's more clear to the audiences that what we don't support.

bh202205 commented on docs/data/sources/databricks.md:
```diff
@@ -0,0 +1,121 @@
+## Limitations
+
+- Databricks import doesn't support data well.
+- This integration supports importing data from Databricks.
```
Is this sentence necessary?

bh202205 commented on docs/data/sources/databricks.md:
```diff
@@ -0,0 +1,121 @@
+
+- Databricks import doesn't support data well.
+- This integration supports importing data from Databricks.
+- User stream view isn't available for data ingested with this feature.
```
Can you point out it's the `User stream view` in `Users&Sessions` in Amplitude so that customers don't confused it with other `User stream view` features? `User stream view` is a quite general term and it can refer to multiple different things.

markzegarelli replied:
@bh202205 So the User Stream view in Users & Sessions isn't available for data ingested with this feature?

bh202205 replied:
That's correct.

bh202205 commented on docs/data/sources/databricks.md:
```diff
@@ -0,0 +1,121 @@
+
+### Authentication
+
+Amplitude's Databricks import supports authentication through a Databricks Workspace User, or Service Principal. Choose Workspace User authentication for faster setup, or Service Principal authentication for finer grained control. 
```
Can we point out they are [Databricks personal access tokens for workspace users](https://docs.databricks.com/en/dev-tools/auth/pat.html#pat-user) and [Databricks personal access tokens for service principals](https://docs.databricks.com/en/dev-tools/auth/pat.html#pat-sp)? Authentication through a Databricks Workspace User, or Service Principal is totally different thing and databricks does support. Let's be more clear here to avoid confusion.
<img width="878" alt="Screenshot 2024-04-01 at 11 13 24 AM" src="https://github.com/amplitude/amplitude-dev-center/assets/105728292/0b583526-4c98-4b72-90a3-4fd2e2cd43ba">
https://docs.databricks.com/en/dev-tools/auth/index.html#databricks-personal-access-tokens-for-workspace-users

bh202205 replied:
I believe it's good to attach this link here as well to give customers a reference to the overview of authentication in databricks: https://docs.databricks.com/en/dev-tools/auth/index.html#authentication-for-databricks-automation---overview

bh202205 commented on docs/data/sources/databricks.md:
```diff
@@ -0,0 +1,121 @@
+
+Amplitude's Databricks import supports authentication through a Databricks Workspace User, or Service Principal. Choose Workspace User authentication for faster setup, or Service Principal authentication for finer grained control. 
+
+#### Get a personal access token (PAT) for the workspace user
```
Can we rename the subtitle to be something like `Databricks personal access tokens for service principals`?

bh202205 commented on docs/data/sources/databricks.md:
```diff
@@ -0,0 +1,121 @@
+
+Amplitude's Databricks import uses Personal Access Tokens to authenticate. For the quickest setup, create a PAT for your workspace user in Databricks. For more information, see Databricks' article [Personal Access Tokens for Workspace Users](https://docs.databricks.com/en/dev-tools/auth/pat.html#databricks-personal-access-tokens-for-workspace-users)
+
+#### Create a service principal (optional)
```
Can we rename the subtitle to be something like `Databricks personal access tokens for service principals`?

bh202205 commented on docs/data/sources/databricks.md:
```diff
@@ -0,0 +1,121 @@
+
+Amplitude recommends that you create a service principal in Databricks to allow for more granular control of access.
+
+1. Follow the Databricks instructions to create a service principal in [Databricks | OAuth machine-to-machine (M2M) authentication](https://docs.databricks.com/en/dev-tools/auth/oauth-m2m.html). Copy the **UUID** for use in a later step.
```
Provide this link as an additional source for our customers to understand how to manage service principals in databricks: https://docs.databricks.com/en/administration-guide/users-groups/service-principals.html

bh202205 commented on docs/data/sources/databricks.md:
```diff
@@ -0,0 +1,121 @@
+Amplitude recommends that you create a service principal in Databricks to allow for more granular control of access.
+
+1. Follow the Databricks instructions to create a service principal in [Databricks | OAuth machine-to-machine (M2M) authentication](https://docs.databricks.com/en/dev-tools/auth/oauth-m2m.html). Copy the **UUID** for use in a later step.
+2. Generate a PAT for the Service Principal
```
Let's add more details and links to help our customers to create personal access tokens for service principals here. They are different from what they need to do for creating personal access tokens of a workspace user.

-----

Generate Personal Access Tokens
Now you can generate a Personal Access Token on this Service Principal to authenticate. Please follow the instructions on the official Databricks documentation [Databricks personal access tokens for service principals](https://docs.databricks.com/en/dev-tools/auth/pat.html#databricks-personal-access-tokens-for-service-principals) if you have an AWS or GCP based databricks. For Azure based databricks, here is the Databricks documentation to follow for generating personal access tokens: [Manage personal access tokens for a service principal](https://learn.microsoft.com/en-us/azure/databricks/administration-guide/users-groups/service-principals#manage-personal-access-tokens-for-a-service-principal).

bh202205 commented on docs/data/sources/databricks.md:
```diff
@@ -0,0 +1,121 @@
+
+| Permission | Reason                                                                               | Location in Databricks                                         |
+| ---------- | ------------------------------------------------------------------------------------ | -------------------------------------------------------------- |
+| Workspace  | Grants access to your Databricks workspace.                                          | *Workspace → <workspace_name> → Permissions → Add permissions* |
```
after `Add permissions`, as pointed out in the [doc](https://docs.google.com/document/d/1_Uri5Czanj8Uhzmzi1jXMhObTlXi-gkUlTDqGJEl1j0/edit), we need to add `Add your created service principal with User permission. And Save`

bh202205 commented on docs/data/sources/databricks.md:
```diff
@@ -0,0 +1,121 @@
+| Permission | Reason                                                                               | Location in Databricks                                         |
+| ---------- | ------------------------------------------------------------------------------------ | -------------------------------------------------------------- |
+| Workspace  | Grants access to your Databricks workspace.                                          | *Workspace → <workspace_name> → Permissions → Add permissions* |
+| Table      | Grants access to list tables and read data.                                          | *Catalog → pick the catalog→ Permissions → Grant*              |
```
Same for granting table permissions. We will need to point out what kind of permission the customers need to grant. In this case, it's `Data Reader` and it's better to list out the detailed permissions as well: `USE CATALOG`, `USE SCHEMA`, `EXECUTE`, `READ VOLUME`, and `SELECT`.

bh202205 commented on docs/data/sources/databricks.md:
```diff
@@ -0,0 +1,121 @@
+| ---------- | ------------------------------------------------------------------------------------ | -------------------------------------------------------------- |
+| Workspace  | Grants access to your Databricks workspace.                                          | *Workspace → <workspace_name> → Permissions → Add permissions* |
+| Table      | Grants access to list tables and read data.                                          | *Catalog → pick the catalog→ Permissions → Grant*              |
+| Cluster    | Grants access to connect to the cluster and run workflows on your behalf             | *Compute → All-purpose compute → Edit Permission*              |
```
After click `Edit Permission`, we will need to let the customers know that they need to add `Add Can Attach To` permission for the service principal.

bh202205 commented on docs/data/sources/databricks.md:
````diff
@@ -0,0 +1,121 @@
+GRANT SELECT ON ANY FILE TO `<service_principal_uuid>`;
+```
+
+### Find or create an all-purpose compute cluster
````
This step should come as the first step in configuration before `Authentication` because without a cluster, customers wouldn't be able to grant necessary permissions for their service principals.

bh202205 commented on docs/data/sources/databricks.md:
```diff
@@ -0,0 +1,121 @@
+
+### Find or create an all-purpose compute cluster
+
+Amplitude creates workflows in this cluster on your behalf to start sync jobs. When complete, copy the **Server hostname** and **Http path** values to use in a later step. Find both values on the **Configuration -> JDBC/ODBC** tab. For more information, see [Databricks | Create a cluster](https://docs.databricks.com/en/dev-tools/auth/oauth-m2m.html)
```
`https://docs.databricks.com/en/dev-tools/auth/oauth-m2m.html` this is the link for service principal. Why do you want to include this link here?

bh202205 replied:
I believe we can include this link here to help customers to understand different compute and what all purpose cluster is in databricks: https://docs.databricks.com/en/compute/index.html

bh202205 reviewed (changes requested):
@markzegarelli Hi Mark, sorry for the delayed comments in this PR. We have finished the most engineering work for databricks import. I will work closely with you now to get this developer doc out. For that, I have made several comments and would you mind addressing them? Thanks!

bh202205 commented:
@markzegarelli Is there any way we can attach a loom video with preview in our developer doc? I have recorded a loom video to explain this point more: https://www.loom.com/share/6b914181c24e4d9bbbefc5bfc85c8a59?sid=325a77b0-518d-49ed-882c-380df10aebf3
And here is the documentation that Mixpanel has for their databricks integration: https://docs.mixpanel.com/changelogs/2023-11-08-databricks-connector

markzegarelli commented:
Thanks @bh202205, I addressed your comments and left a clarifying question of my own.

I'd prefer not to have a loom video in the docs, as we don't include them in other docs, and I don't want to set a precedence. If there is a big demand for it, let's discuss how we can best maintain this sort of content.

bh202205 commented:
> Thanks @bh202205, I addressed your comments and left a clarifying question of my own.
> 
> I'd prefer not to have a loom video in the docs, as we don't include them in other docs, and I don't want to set a precedence. If there is a big demand for it, let's discuss how we can best maintain this sort of content.

@markzegarelli As we discussed in [the Slack thread](https://amplitude.slack.com/archives/C0579CK1936/p1712003737451249?thread_ts=1711997487.030379&cid=C0579CK1936), can you include the following loom video link in the doc? Thanks!
https://www.loom.com/share/b287dce8c48145bd845c7ea1122b20c4?sid=5a4661f2-9d27-4e32-88c8-582155f12d5d

bh202205 commented:
> Thanks @bh202205, I addressed your comments and left a clarifying question of my own.
> 
> I'd prefer not to have a loom video in the docs, as we don't include them in other docs, and I don't want to set a precedence. If there is a big demand for it, let's discuss how we can best maintain this sort of content.

@markzegarelli as we discussed [in the Slack thread](https://amplitude.slack.com/archives/C0579CK1936/p1712003737451249?thread_ts=1711997487.030379&cid=C0579CK1936), can you include the following loom video in to doc? Thanks!
https://www.loom.com/share/a00f8905170e4c83977ae6fb2f0dcde7?sid=5a77e8c9-d34b-42b0-a179-679669c8bdbe

bh202205 reviewed (approved):
@markzegarelli Thank you for collaborate with me on this! The doc looks all good now. Let's get this published!

fzqgriffin commented on docs/data/sources/databricks.md:
```diff
@@ -0,0 +1,125 @@
+
+## Limitations
+
+- Databricks import doesn't support export data to your own destinations.
```
@bh202205 this is probably not true anymore since we are able to export data ingested through cargo. @LeontiBrechko can help to confirm?

LeontiBrechko replied:
Cargo-ingested data has support of export to DWH and CS destinations

bh202205 replied:
https://github.com/amplitude/amplitude-dev-center/pull/275/files#r1546944408

fzqgriffin commented on docs/data/sources/databricks.md:
```diff
@@ -0,0 +1,125 @@
+
+- Databricks import doesn't support export data to your own destinations.
+- User stream view isn't available for data ingested with this feature.
+- End-to-end latency (from when Amplitude fetches the data, to when you can query it) is between 30 and 60 minutes. Actual latency can vary depending on the workload.
```
this statement is not accurate anymore since V2 implementation is much faster than V1 implementation, now, usually the latency is 15mins(mostly due to scheduling overheads in falcon, which is the same for all integrations).

Also, felt like we should remove this statement since it can kind of causing impression that the feature will be much slower than other integrations, which is not correct.

bh202205 replied:
Updated in the new PR: https://github.com/amplitude/amplitude-dev-center/pull/275/

fzqgriffin commented on docs/data/sources/databricks.md:
```diff
@@ -0,0 +1,125 @@
+## Limitations
+
+- Databricks import doesn't support export data to your own destinations.
+- User stream view isn't available for data ingested with this feature.
```
hey @shawn-amplitude do you know if this is still true?

shawn-amplitude replied:
We need to define `user stream view` here.
- We can't see users in this page: https://app.amplitude.com/analytics/amplitude/activity
- But if we have the user id, we can see events here: https://app.amplitude.com/analytics/amplitude/project/506317/search/amplitude_id%3D862380164641?collapsed=true&sessionHandle=%2524jsn%255B%2522-B%2522%252C%2522Y6HkFOq%2522%252C%2522MjJ3QYh%2522%252C%2522B7nN%2522%252C%2522device_id_1711671366570_68%2522%255D&eventId=01752a7b-42b1-41af-bea4-287097b92331

fzqgriffin replied:
@shawn-amplitude yep, the `User Look Up` page definitely is not showing the most recent 100 events there, which is not expected. Let's probably reword the statement to `User Look-Up page won't be able to display most recent 100 events ingested through this feature yet`.

Need to ask if anyone knows how that feature is implemented and confirm with izaaz to see if this is the issue he referred to.

bh202205 replied:
https://github.com/amplitude/amplitude-dev-center/pull/275/files#r1546945076

fzqgriffin commented on docs/data/sources/databricks.md:
```diff
@@ -0,0 +1,125 @@
+
+### Find or create an all-purpose compute cluster
+
+Amplitude creates workflows in this cluster on your behalf to start sync jobs. When complete, copy the **Server hostname** and **Http path** values to use in a later step. Find both values on the **Configuration -> JDBC/ODBC** tab. For more information about cluster types, see [Compute](https://docs.databricks.com/en/compute/index.html).
```
Let's also add instruction like below to ensure cluster has right permissions.

```
Ensure that the new cluster is able to run jobs by NOT having configs below in cluster's policy. See details in [doc](https://docs.databricks.com/en/administration-guide/clusters/policy-definition.html#workload).

 "workload_type.clients.jobs": {
    "type": "fixed",
    "value": false
  }
```

bh202205 replied:
Updated in the new PR: https://github.com/amplitude/amplitude-dev-center/pull/275/
